Document we're not tracking relationships between symbols (#16018)
python/mypy: docs/source/type_narrowing.rst

The simplest way to narrow a type is to use one of the supported expressions:

- :py:func:`isinstance` like in :code:`isinstance(obj, float)` will narrow ``obj`` to have ``float`` type
- :py:func:`issubclass` like in :code:`issubclass(cls, MyClass)` will narrow ``cls`` to be ``Type[MyClass]``
- :py:class:`type` like in :code:`type(obj) is int` will narrow ``obj`` to have ``int`` type
- :py:func:`callable` like in :code:`callable(obj)` will narrow object to callable type

> - :code:`obj is not None` will narrow object to its :ref:`non-optional form <strict_optional>`

Author: strangely we didn't mention this

     reveal_type(x)  # Revealed type is "builtins.int"
     print(x + '!')  # Typechecks with `mypy`, but fails in runtime.

Limitations of narrowing

> ~~~~~~~~~~~~~~~~~~~~~~~~

Author: Hope this is the right place. I wanted it to have a heading, to make it easy to refer people, but it's also a bit strange that it comes as a topic in the same hierarchy and right before "is_subclass" and "callable".

Reviewer: Let's make it a heading and move it to the bottom of the page.

Mypy's analysis is limited to individual symbols and it will not track
relationships between symbols. For example, in the following code
it's easy to deduce that if :code:`a` is None then :code:`b` must not be,
therefore :code:`a or b` will always be a string, but Mypy will not be able to tell that:

.. code-block:: python

    def f(a: str | None, b: str | None) -> str:
        if a is not None or b is not None:
            return a or b  # Incompatible return value type (got "str | None", expected "str")
        return 'spam'

Tracking these sort of cross-variable conditions in a type checker would add significant complexity

> and performance overhead and would be computationally infeasible in all but the most basic cases.

Reviewer: ```suggestion
and performance overhead.
```
Claim feels a little too strong, first half of the sentence says enough.

> You may override the type checker with a :ref:`cast <casts>`, or rewrite the function to be

Reviewer: ```suggestion
You may override the type checker with an ``assert`` or :ref:`cast <casts>`, or rewrite the function to be
```
When I run into this, assert is usually more ergonomic. It's also safer than cast.

Author: Probably would phrase as "help the type checker with an ``assert``", but I'm not sure it'll be more ergonomic in this case. Maybe I'm tired and missing something obvious?

cast:
```py
def f(a: str | None, b: str | None) -> str:
    if a is not None or b is not None:
        return a or cast(str, b)
    return 'spam'
```

assert:
```py
def f(a: str | None, b: str | None) -> str:
    if a is not None or b is not None:
        if a:
		   return a
		assert b
		return b
    return 'spam'
```

third option:
```py
def f(a: str | None, b: str | None) -> str:
    if a is not None or b is not None:
        return a or b or 'impossible'
    return 'spam'
```

Reviewer: Yeah, not more ergonomic in this case, but better in the cases where I usually run into this. But maybe I dislike cast more than most people :-)

Author: `assert a or b` doesn't help you either because we don't narrow conditionally. (Also, in this case you'd get more than you'd bargain for, as empty `b` would assert.)

Reviewer: Oh oops, I guess you'd have to do `ret = a or b; assert ret is not None; return ret`

.. note::

  The same will work with ``isinstance(x := a, float)`` as well.

Limitations
-----------

Mypy's analysis is limited to individual symbols and it will not track
relationships between symbols. For example, in the following code
it's easy to deduce that if :code:`a` is None then :code:`b` must not be,
therefore :code:`a or b` will always be a string, but Mypy will not be able to tell that:

.. code-block:: python

    def f(a: str | None, b: str | None) -> str:
        if a is not None or b is not None:
            return a or b  # Incompatible return value type (got "str | None", expected "str")
        return 'spam'

Tracking these sort of cross-variable conditions in a type checker would add significant complexity
and performance overhead.

> You may override the type checker with a :ref:`cast <casts>`, help it with an assertion,

Reviewer: ```suggestion
You can use an ``assert`` to convince the type checker, override it with a :ref:`cast <casts>`
```
assert is safer than cast, so it should be first option
